### ISSUE TITLE
Directories in ISO Charset not shown in Project explorer/Java 19 issue

### ISSUE THREAD

**Issue by mrhaarmann** [eclipse-platform/eclipse.platform.ui#1060]
We are running Eclipse IDE version 2023-06 in two different environments, since one older project requires ISO-8859-15 locale.
(in Linux version)
In order to achieve this, we are starting eclipse with a number of environment settings:
export LANG=de_DE@euro
export LANGUAGE=de_DE:de:en_GB:en
export LANG=de_DE@euro
export LC_ALL=de_DE@euro
export GDM_LANG=de_DE@euro
export LC_NAME=de_DE@euro
export LC_MONETARY=de_DE@euro
export LC_MEASUREMENT=de_DE@euro
export LC_TELEPHONE=de_DE@euro
export LC_PAPER=de_DE@euro
export LC_IDENTIFICATION=de_DE@euro
export LC_TIME=de_DE@euro
export LC_NUMERIC=de_DE@euro
export LC_ADDRESS=de_DE@euro
This will lead to filenames with ISO-8859-15 encoding and also content in ISO-8859-15 encoding, when the workspace is set to encoding from environment.
We recently tried to execute Eclipse with Java 19 instead of the Java 17 version and found an issue which might be addressed by the platform.
When eclipse is started in 8859-15 locale, directory names which contain german Umlaut characters (such as äöü) are not displayed any more in the Project explorer (when refreshing the list). Even new directories are created correctly (on the filesystem, everything looks hood) and vanish from view when a refresh is done.
This can also be reproduced with Eclipse 2023-03, running with Java 19.
The JDT people told me they rely on the underlying platform code, so thats why I open an issue here.
For the meantime, we are running with Java 17, but at some point in the future, an upgrade will be needed.

**Comment by merks** [comment]
Could you please attach a zip file of the exported sample project?  Use File -> Export... -> General -> Archive File and include everything in the zip, especially also any hidden .* files in the root, e.g., the .project file.  Then you can drag and drop the *.zip onto the issue comment area to attach it.

**Comment by laeubi** [comment]
> This will lead to filenames with ISO-8859-15 encoding and also content in ISO-8859-15 encoding, when the workspace is set to encoding from environment.

have you tried to set the workspace encoding instead?

**Comment by mrhaarmann** [comment]
Setting the workspace encoding does not help, however the specific locale is not available there (only ISO-8859-1).
The export does NOT produce the problematic directory at all.
I did a manual zip in a terminal with ISO-8859-15 locale now.
[test.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/12415828/test.zip)

**Comment by laeubi** [comment]
> Setting the workspace encoding does not help, however the specific locale is not available there (only ISO-8859-1).

Okay, that seems to indicate that `ISO-8859-15` locale is simply not available to the JVM, so I fear there is not much one can do (beside from migrating to `ISO-8859-1` or `UTF-8`), one can probably create a simple Java program that checks what locales are returned from a Java 17 versus a Java 19 JVM....

I just found here how to implement a custom charset that seem to reference  `ISO-8859-15` / `IBM_924`
https://stackoverflow.com/questions/76936549/charset-not-present-in-used-jdk/76936879#76936879 so there *might* be some way to add support for that into eclipse with some similar code even if the JVM does not support it (anymore).

**Comment by merks** [comment]
Even my OS (Wndows 10) does not display this file correctly:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/42dc42cf-9222-4794-8bda-f23d19aa24ef)

**Comment by laeubi** [comment]
By the way this indicates Java 19 *should* support this:

https://docs.oracle.com/en/java/javase/19/intl/supported-encodings.html#GUID-E20951E6-C420-4D2F-A6BE-1470B4D55B3B

**but** it seem not always working from Strings:

@merks encodings in Zip files is another story, I don't think its really possible to transfer such encoded names reliable via Zip, you most likely need to use ZipFile with a specified encoding, but this will extract the file in **your** encoding.

**Long Story Short:** Using special encodings for filenames is waiting for problems to happen if your system does not really use that encoding.

**Comment by merks** [comment]
Yes, one should avoid this like one would avoid the plague.  I don't even quite grasp how a file system on an OS could deal with such things; surely the names of files in the file system must be encoded in such a way that it works properly/correctly regardless of environment variable settings.  I thought Windows itself used UTF-16 to encode file names...

**Comment by laeubi** [comment]
It all depends a bit, as you can see your OS can *store* this special character, it just looks *odd* to you and probably there are some tools that might jerk when opening the file.

Nerveless it seems to me that Eclipse can't do much (if its just the Java version that differs here) @mrhaarmann any chance you can look at the eclipse error log if you see something there indicating an exception or similar?

**Comment by iloveeclipse** [comment]
> When eclipse is started in 8859-15 locale, directory names which contain german Umlaut characters (such as äöü) are not displayed any more in the Project explorer (when refreshing the list)

I don't see any encoding set in the attached project .settings. 
There should be `.settings/org.eclipse.core.resources.prefs` file containing 
```
eclipse.preferences.version=1
encoding/<project>=ISO-8859-15
```

But this isn't the problem here (because it is not considered for file **names** encoding).
I see the place we've missed during https://openjdk.org/jeps/400 [adoption](https://github.com/eclipse-platform/eclipse.platform/pull/38). 
I **believe** we need an update of https://github.com/eclipse-platform/eclipse.platform/blob/adca2ac4e507072fddf06d0da1e5176a8661bf3b/resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/local/Convert.java#L23. Currently this returns **charset** used for **file content** encoding but we need system charset used for **file names**. Changing it to `System.getProperty("native.encoding")` (which is what Platform does in [getSystemCharset()](https://github.com/eclipse-platform/eclipse.platform/blob/adca2ac4e507072fddf06d0da1e5176a8661bf3b/runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Platform.java#L1513)) fixes the problem in debugger:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/964108/fa137627-c277-44cd-9928-d71a3cf5f266)

```
diff --git a/resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/local/Convert.java b/resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/local/Convert.java
index 4ac7f0a..b653cf8 100644
--- a/resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/local/Convert.java
+++ b/resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/local/Convert.java
@@ -22,3 +22,3 @@
 	/** Indicates the default string encoding on this platform */
-	private static String defaultEncoding = new java.io.InputStreamReader(new java.io.ByteArrayInputStream(new byte[0])).getEncoding();
+	private static String defaultEncoding = System.getProperty("native.encoding"); //$NON-NLS-1$
```
